### PR TITLE
[appdomain] Catch exceptions when invoking AppDomain.DoAssemblyResolve

### PR DESCRIPTION
**Pull request by lambdageek**
Otherwise we may leak a coop handle for the requesting assembly, which may
lead to an assertion when we unload the current domain.

Fixes [#58950](https://bugzilla.xamarin.com/show_bug.cgi?id=58950)

### CODE REVIEW

**Comment by lambdageek** [comment]
Added a regression test

File: mono/tests/assemblyresolve_event6.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		public int DoSomething ()
+		{
+			AppDomain.CurrentDomain.ReflectionOnlyAssemblyResolve += new ResolveEventHandler (MyReflectionResolveEventHandler);
+			bool caught = false;
+			try {
+				Assembly a = Assembly.ReflectionOnlyLoadFrom ("assemblyresolve_asm.dll");
+				var t = a.GetType ("Asm2");
+				var m = t.GetMethod ("M"); // This triggers a load of TestBase
+				Console.WriteLine ("got '{0}'", m);
+			} catch (FileNotFoundException e) {
```
Comment (kumpera) [inline mono/tests/assemblyresolve_event6.cs]: Which one is supposed to be thrown? FNFE or MyExn?

**Comment by kumpera** [comment]
LGTM other than the regression test

File: mono/tests/assemblyresolve_event6.cs (hunk 1 above)
Comment (lambdageek) [reply mono/tests/assemblyresolve_event6.cs]: oops. thanks - should be FNFE - MyExn would be the leaky previous behavior.

**Comment by lambdageek** [comment]
Fixed the test; Win 64 debugger failure seems unrelated.

**Pull request merged**